### ISSUE TITLE
2.5.1+ doesn't initialize when used with azure-functions-core-tools

### ISSUE THREAD

**Issue by CreativeTechGuy** [microsoft/ApplicationInsights-node.js#1144]
TLDR: Two different libraries, `applicationinsights` and `azure-functions-core-tools` both overwrite Node.js' `require` method, thus breaking the magic that `azure-functions-core-tools` is doing and preventing `applicationinsights` to hook into the local Azure Function server. This causes automatic logging when running Azure Functions locally to not work and `getCorrelationContext()` to return null.

This is an incredibly complicated issue. I feel it's best explained by tracing the code since there's a lot going on.

* The offending commit in this repo is https://github.com/microsoft/ApplicationInsights-node.js/commit/19d92af2501587459885c7b097848ed0c01f37ec. This added a dependency on `@azure/opentelemetry-instrumentation-azure-sdk`.
* That dependency depends on [@opentelemetry/instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation)
* Which that dependency depends on [require-in-the-middle](https://www.npmjs.com/package/require-in-the-middle), a dependency which overwrites the `require` method in Node.js. Make a note of this for later.
* For `applicationinsights` to add the proper Azure Functions hooks, it checks if `@azure/functions-core` exists [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/45e06c608a291442ad2a100dee771fc1e8908d74/AutoCollection/AzureFunctionsHook.ts#LL22C13-L29C14)
* This `@azure/functions-core` package doesn't actually exist in NPM or on disk, it is instead provided by [azure-functions-core-tools](https://www.npmjs.com/package/azure-functions-core-tools) when developing locally. It does this by overwriting the `require` method in Node.js such that when that specific module is requested, it provides an in-memory object as the response.
* When installing `azure-functions-core-tools` it will make a request to install [this file](https://functionscdn.azureedge.net/public/4.0.5148/Azure.Functions.Cli.win-x64.4.0.5148.zip) as a postinstall script and unpack it into `node_modules`. The source relevant source code which that zip includes is for the [azure-functions-nodejs-worker](https://github.com/Azure/azure-functions-nodejs-worker) package.
* In that worker package, [here's the code](https://github.com/Azure/azure-functions-nodejs-worker/blob/c919f101d61acc9a4bda49bbbc035dc585bd244a/src/setupCoreModule.ts#L48-L56) which overwrites `require` to supply the fake `@azure/functions-core` package.

Now, the problem is that there's two packages which are loaded at runtime, both overwriting a global with a custom implementation. `azure-functions-core-tools` is loaded first so it overwrites first, then `require-in-the-middle` loads second and overwrites the overwrite. Thus when `applicationinsights` tries to check for the existence of `@azure/functions-core` at runtime, it cannot find it. As a result of all of this. our calls to `appinsights.getCorrelationContext()` return null as it was never initialized as the Azure Function hooks were never setup.

This issue was introduced in 2.5.1 and was not an issue in 2.5.0 of `applicationinsights`. It's unclear which package is the root cause, it seems like both are equally to blame as they aren't playing nice with each other. Both `azure-functions-core-tools` and `require-in-the-middle` "try" to overwrite in a way to be compatible, but due to the use of the Proxy in `azure-functions-core-tools` it is still incompatible if loaded first.

**Comment by JacksonWeber** [comment]
Hi @CreativeTechGuy, if you want to eliminate the conflict here you could disable Azure SDK auto-instrumentation. This won't fix the core issue that the two dependencies both monkey-patch the require method but might be able to resolve your issue for now. Let me know that workaround helps, thanks!

**Comment by CreativeTechGuy** [comment]
@JacksonWeber This doesn't change anything unfortunately. The issue is that the auto-instrumentation doesn't work. So manually disabling the thing that doesn't work doesn't really do anything.

**Comment by ejizba** [comment]
I took a look and it seems like we're just waiting for the fix to propogate through various dependencies.

1. `require-in-the-middle` [fixed this](https://github.com/elastic/require-in-the-middle/pull/59) in [v7.0.0](https://github.com/elastic/require-in-the-middle/blob/main/CHANGELOG.md#v700)
2. `@opentelemetry/instrumentation` [updated](https://github.com/open-telemetry/opentelemetry-js/pull/3727) to `require-in-the-middle` v7.0.0 in their [v0.39.0 release](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/CHANGELOG.md#0390)
3. `@azure/opentelemetry-instrumentation-azure-sdk` is [currently on](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/instrumentation/opentelemetry-instrumentation-azure-sdk/package.json#L84) `@opentelemetry/instrumentation` v0.38.0 and needs to update to v0.39.0
4. `applicationinsights` will need to update to the latest `@azure/opentelemetry-instrumentation-azure-sdk`

@JacksonWeber @hectorhdzg if you could help speed along those last two updates of packages that would be much appreciated :)

**Comment by CreativeTechGuy** [comment]
Thank you @ejizba!

**Comment by JacksonWeber** [comment]
@ejizba Absolutely. Thank you for noticing the updates in the dependency web there!

**Comment by JacksonWeber** [comment]
@CreativeTechGuy @ejizba The above issue should be resolved as our dependency on `azure/opentelemetry-instrumentation` has been updated and that package's dependence on `require-in-the-middle` and all opentelemetry packages has been updated.

**Comment by CreativeTechGuy** [comment]
Is there a plan to release that fix? It looks like it's been committed but the package hasn't been released in a few weeks.

**Comment by hectorhdzg** [comment]
@CreativeTechGuy we have a monthly release cadence in our side, we will be publishing a new version of Application Insights Node.js SDK end of next week.
